internal/compile/compile.go

```diff
@@ -68,36 +67,22 @@ func ExtractUnknownsFromAnnotations(comp *ast.Compiler, ref ast.Ref) ([]*ast.Ter
 	return unknownsFromAnnotationsSet(comp.GetAnnotationSet(), rule)
 }
 
-func unknownsFromAnnotationsSet(as *ast.AnnotationSet, rule *ast.Rule) ([]*ast.Term, []*ast.Error) {
+func unknownsFromAnnotationsSet(as *ast.AnnotationSet, rule *ast.Rule) ([]ast.Ref, []*ast.Error) {
 	if as == nil {
 		return nil, nil
 	}
-	var unknowns []*ast.Term
+	var unknowns []ast.Ref
 	var errs []*ast.Error
 
 	for _, ar := range as.Chain(rule) {
 		ann := ar.Annotations
-		if ann == nil {
+		if ann == nil || ann.Compile == nil {
 			continue
 		}
-		unk, ok := ann.Custom["unknowns"]
-		if !ok {
-			continue
-		}
-		unkArray, ok := unk.([]any)
-		if !ok {
-			continue
-		}
-		for _, u := range unkArray {
-			s, ok := u.(string)
-			if !ok {
-				continue
-			}
-			ref, err := ast.ParseRef(s)
-			if err != nil {
-				errs = append(errs, ast.NewError(invalidUnknownCode, ann.Loc(), "unknowns must be valid refs: %s", s))
-			} else if ref.HasPrefix(ast.DefaultRootRef) || ref.HasPrefix(ast.InputRootRef) {
-				unknowns = append(unknowns, ast.NewTerm(ref))
+		unkArray := ann.Compile.Unknowns
+		for _, ref := range unkArray {
+			if ref.HasPrefix(ast.DefaultRootRef) || ref.HasPrefix(ast.InputRootRef) {
+				unknowns = append(unknowns, ref)
```

Comment:
We could end up with duplicates in the array here, would that have any consequences for e.g. error reporting?